<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-задачи" data-toc-modified-id="Описание-задачи-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание задачи</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Ответы-на-задания:" data-toc-modified-id="Ответы-на-задания:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на задания:</a></span><ul class="toc-item"><li><span><a href="#Посчитаем-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитаем сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку;-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключите-из-анализа-брошюры" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключите из анализа брошюры</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывая-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывая-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li></ul></li></ul></div>

## Описание задачи

**Цель** - проанализировать базу данных купленного сервиса для чтения книг по подписке для дальнейшего формулирования ценностного предложения для нового продукта.

**Задачи:**

познакомиться с хранящимися в базе данными,

проанализировать результаты запросов.

Данные расположены в 5 таблицах, содержащих информацию о книгах, издательствах, авторах, пользовательских оценках и обзорах книг.

## Описание данных

Таблица `books` - содержит данные о книгах:

book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

Таблица `authors` - содержит данные об авторах:

author_id — идентификатор автора;

author — имя автора.

Таблица `publishers` - содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

Таблица `ratings` - содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

Таблица `reviews` - содержит данные о пользовательских обзорах на книги:  

review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя пользователя, написавшего обзор;

text — текст обзора.

In [1]:
!pip install sqlalchemy
!pip install psycopg2-binary

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
query = \
'''
SELECT * 
FROM authors LIMIT 5
'''

pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
query = \
'''
SELECT * 
FROM publishers LIMIT 5
'''

pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
query = \
'''
SELECT * 
FROM ratings LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
query = \
'''
SELECT *
FROM reviews LIMIT 5
'''

pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:
query = \
'''
SELECT *
FROM reviews LIMIT 5
'''
'''
SELECT * 
FROM ratings LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Ответы на задания:

### Посчитаем сколько книг вышло после 1 января 2000 года

In [8]:
query = \
'''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года вышло 819 книг

### Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [9]:
query = ''' SELECT title, SUM(cnt_review) AS cnt_review, AVG(avg_raiting) AS avg_raiting
            FROM books 
            LEFT JOIN (SELECT book_id, COUNT(review_id) AS cnt_review 
                        FROM reviews
                        GROUP BY book_id) AS sub
            ON books.book_id = sub.book_id
            LEFT JOIN (SELECT book_id, AVG(rating) AS avg_raiting
                        FROM ratings 
                        GROUP BY book_id) AS sub_2
            ON books.book_id = sub_2.book_id
            GROUP BY books.book_id
        '''
pd.io.sql.read_sql(query, con = engine)

title  cnt_review  \
0            The Body in the Library (Miss Marple  #3)         2.0   
1                                            Galápagos         2.0   
2                             A Tree Grows in Brooklyn         5.0   
3    Undaunted Courage: The Pioneering First Missio...         2.0   
4                    American Gods (American Gods  #1)         5.0   
..                                                 ...         ...   
995                                         Possession         3.0   
996          A Woman of Substance (Emma Harte Saga #1)         2.0   
997                                          Christine         3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999                           The Plot Against America         2.0   

     avg_raiting  
0       4.500000  
1       4.500000  
2       4.250000  
3       4.000000  
4       4.000000  
..           ...  
995     4.000000  
996     5.000000  
997     3.428571  
998     3.500000  
999     3.000000  

[1000 rows x 3 columns]

Из таблицы видно, что для большинства книг было написано несколько обзоров и средний рейтинг варьируются.

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключите из анализа брошюры

In [10]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. 

### Определим автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками

In [11]:
query = '''
SELECT
    authors.author AS автор,
    AVG(ratings.rating) AS средний_рейтинг
    FROM ratings
    JOIN books ON books.book_id = ratings.book_id
    JOIN authors ON books.author_id = authors.author_id
    WHERE ratings.book_id IN (
       SELECT ratings.book_id
       FROM ratings
       GROUP BY ratings.book_id
       HAVING COUNT(ratings.book_id) >= 50)
    GROUP BY authors.author_id
    ORDER BY AVG(ratings.rating) DESC
    LIMIT 1'''
pd.io.sql.read_sql(query, con = engine)

автор  средний_рейтинг
0  J.K. Rowling/Mary GrandPré         4.287097

Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Поттере. Это действительно мировой бестселлер вышедший рекордным тиражом.

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [12]:
query = '''

  SELECT AVG (reviews_count) as avg_reviews_count
  FROM ( 
        SELECT username,
        COUNT(review_id) as reviews_count
        FROM reviews
        GROUP BY username
        HAVING username IN (
                            SELECT username
                            FROM ratings
                            GROUP BY username
                            HAVING COUNT (rating_id) > 48
                            )
                            
                            
        ) as reviews
  
  '''
pd.io.sql.read_sql(query, con = engine)

avg_reviews_count
0               24.0

Читатели, которые поставили больше 48 оценок, в среднем пишут по 24 обзора

**Вывод:**

В нашем распоряжении было 5 таблиц:

books - содержит 1000 строк и 6 столбцов, содержит данные о книгах;
authors - содержит 636 строк и 2 столбца, содержит данные об авторах;
publishers - содержит 340 строк, 2 столбца, содержит данные об издательствах;
rating - содержит 6456 строк, 4 столбца, содержит данные о пользовательских обзорах;
reviews - содержит 2793 строки, 4 столбца, содержит данные о пользовательских обзорах.
В ходе работы были даны ответы на 5 поставленных вопросов:

Cколько книг вышло после 1 января 2000 года? с 1 января 2000 года, в базе хранится информация о 819 книгах

Для каждой книги посчитайте количество обзоров и среднюю оценку: из таблицы видно, что для большинства книг было написано несколько обзоров и средний рейтинг варьируются.

Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — таким образом исключим из анализа брошюры. Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. 

Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками. Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Поттере. Это действительно мировой бестселлер вышедший рекордным тиражом.

Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.Таким образом среднее количество обзоров от активных пользователей. При подсчете было выявлено - 24 обзора.

1. Выведите таблицу, которая будет содержать по году публикации:
количество издательств,
выпущенных книг и
сколько всего тысяч страниц было в изданных книгах
(отобразить только те года, в которых издано более 30 книг)

In [13]:
query = '''

SELECT
    EXTRACT(YEAR FROM b.publication_date) AS publication_year,
    COUNT(DISTINCT b.publisher_id) AS num_publishers,
    COUNT(*) AS num_books,
    SUM(b.num_pages) / 1000.0 AS total_pages_in_thousands
FROM
    books AS b
GROUP BY
    publication_year
HAVING
    COUNT(*) > 30
ORDER BY
    publication_year;
'''
pd.io.sql.read_sql(query, con = engine)

publication_year  num_publishers  num_books  total_pages_in_thousands
0            1999.0              26         41                    15.763
1            2000.0              35         38                    13.328
2            2001.0              41         60                    21.758
3            2002.0              62         94                    38.597
4            2003.0              65        105                    41.423
5            2004.0              88        124                    46.779
6            2005.0              89        139                    55.967
7            2006.0             109        184                    68.302
8            2007.0              38         50                    18.258

2. Выведите в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделайте выводы какой рейтинг больше

In [14]:
query = '''
SELECT
    AVG(r.rating) AS average_rating
FROM
    ratings AS r
WHERE
    r.book_id NOT IN (
        SELECT
            r2.book_id
        FROM
            ratings AS r2
        GROUP BY
            r2.book_id
        HAVING
            COUNT(DISTINCT r2.username) > 3)
UNION

SELECT
    AVG(r.rating) AS average_rating
FROM
    ratings AS r
WHERE
    r.book_id IN (
        SELECT
            r2.book_id
        FROM
            ratings AS r2
        GROUP BY
            r2.book_id
        HAVING
            COUNT(DISTINCT r2.username) > 3)
    
'''
pd.io.sql.read_sql(query, con = engine)

average_rating
0        3.890909
1        3.938398

3 Выведите топ семь пользователей по суммарному показателю написанных ревью и поставленных оценок, указывая суммарное количество страниц в книгах, на которые они отреагировали

In [21]:
query = '''

SELECT username,
       total_reviews+count_scores AS total_sum,
       total_pages_reviewed
FROM
(SELECT rat.username,
       COUNT(rev.review_id) AS total_reviews,
       COUNT(rat.rating_id) AS count_scores,
       SUM(num_pages) AS total_pages_reviewed
FROM BOOKS AS b
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
LEFT JOIN reviews AS rev ON rat.book_id = rev.book_id AND rat.username = rev.username 

GROUP BY rat.username) AS my_table
ORDER BY total_sum DESC
LIMIT 7
'''
pd.io.sql.read_sql(query, con = engine)

username  total_sum  total_pages_reviewed
0      martinadam         83                 21213
1     sfitzgerald         83                 20938
2       richard89         81                 21013
3          paul88         78                 22060
4  jennifermiller         78                 20932
5         susan85         78                 19266
6     lesliegibbs         76                 18113